// @aws-sdk/client-eventbridge/dist-es/commands/PutEventsCommand.js
import { getInjectEndpointIdPlugin } from "@aws-sdk/middleware-sdk-eventbridge";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutEventsRequestFilterSensitiveLog, PutEventsResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1PutEventsCommand, serializeAws_json1_1PutEventsCommand } from "../protocols/Aws_json1_1";
export class PutEventsCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getInjectEndpointIdPlugin(configuration));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EventBridgeClient";
        const commandName = "PutEventsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: PutEventsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: PutEventsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_json1_1PutEventsCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_json1_1PutEventsCommand(output, context);
    }
}
