// @aws-sdk/client-eventbridge/dist-es/commands/PutTargetsCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutTargetsRequestFilterSensitiveLog, PutTargetsResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_json1_1PutTargetsCommand, serializeAws_json1_1PutTargetsCommand, } from "../protocols/Aws_json1_1";
export class PutTargetsCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EventBridgeClient";
        const commandName = "PutTargetsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: PutTargetsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: PutTargetsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_json1_1PutTargetsCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_json1_1PutTargetsCommand(output, context);
    }
}
