// @aws-sdk/client-eventbridge/dist-es/commands/RemovePermissionCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { RemovePermissionRequestFilterSensitiveLog } from "../models/models_0";
import { deserializeAws_json1_1RemovePermissionCommand, serializeAws_json1_1RemovePermissionCommand, } from "../protocols/Aws_json1_1";
export class RemovePermissionCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EventBridgeClient";
        const commandName = "RemovePermissionCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: RemovePermissionRequestFilterSensitiveLog,
            outputFilterSensitiveLog: (output) => output,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_json1_1RemovePermissionCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_json1_1RemovePermissionCommand(output, context);
    }
}
