// @pristine-ts/aws/dist/lib/esm/clients/sqs.client.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { inject, injectable } from "tsyringe";
import { SQSClient, SendMessageCommand } from "@aws-sdk/client-sqs";
import { SqsSendMessageError } from "../errors/sqs-send-message.error";
import { moduleScoped, tag } from "@pristine-ts/common";
import { AwsModuleKeyname } from "../aws.module.keyname";
/**
 * The client to use to interact with AWS SQS. It is a wrapper around the SQSClient of @aws-sdk/client-sqs.
 * It is tagged so it can be injected using SqsClientInterface.
 */
let SqsClient = class SqsClient {
    /**
     * The client to use to interact with AWS SQS. It is a wrapper around the SQSClient of @aws-sdk/client-sqs.
     * @param logHandler The log handler used to output logs.
     * @param region The aws region for which the client will be used.
     */
    constructor(logHandler, region) {
        this.logHandler = logHandler;
        this.region = region;
    }
    /**
     * Returns the instantiated SQSClient from the @aws-sdk/client-sqs library.
     * @param endpoint The endpoint for which the SQS client is created.
     */
    getClient(endpoint) {
        return new SQSClient({
            region: this.region,
            endpoint: endpoint !== null && endpoint !== void 0 ? endpoint : undefined,
        });
    }
    /**
     * Sends a message to the specified Queue URL.
     * @param queueUrl The queue url where to send the message.
     * @param body The body of the message to send in the queue.
     * @param messageGroupId The message group id for FIFO queues.
     * @param delaySeconds The length of time, in seconds, for which to delay a specific message.
     * @param endpoint The endpoint for SQS.
     */
    send(queueUrl, body, messageGroupId, delaySeconds, endpoint) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const client = this.getClient(endpoint);
                const command = new SendMessageCommand({
                    QueueUrl: queueUrl,
                    MessageBody: body,
                    MessageGroupId: messageGroupId,
                    DelaySeconds: delaySeconds,
                });
                this.logHandler.debug("Sending a message to the queue", {
                    queueUrl,
                    body,
                    messageGroupId,
                    delaySeconds,
                }, AwsModuleKeyname);
                const response = yield client.send(command);
                this.logHandler.debug("Message successfully sent to the queue", {
                    queueUrl,
                    body,
                    messageGroupId,
                    delaySeconds,
                    response,
                }, AwsModuleKeyname);
                return {
                    messageId: response.MessageId,
                };
            }
            catch (error) {
                this.logHandler.error("There was an error sending the message to the queue", {
                    error,
                    queueUrl,
                    body,
                    messageGroupId,
                    delaySeconds,
                }, AwsModuleKeyname);
                throw new SqsSendMessageError(error, queueUrl, body, messageGroupId, delaySeconds);
            }
        });
    }
};
SqsClient = __decorate([
    tag("SqsClientInterface"),
    moduleScoped(AwsModuleKeyname),
    injectable(),
    __param(0, inject("LogHandlerInterface")),
    __param(1, inject("%pristine.aws.region%")),
    __metadata("design:paramtypes", [Object, String])
], SqsClient);
export { SqsClient };
