// @pristine-ts/networking/dist/lib/esm/mappers/request.mapper.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { moduleScoped, Request, Response, ServiceDefinitionTagEnum, tag } from "@pristine-ts/common";
import { injectable } from "tsyringe";
import { NetworkingModuleKeyname } from "../networking.module.keyname";
import { Event } from "@pristine-ts/core";
let RequestMapper = class RequestMapper {
    supportsMapping(rawEvent, executionContext) {
        return rawEvent instanceof Request;
    }
    map(rawEvent, executionContext) {
        return {
            executionOrder: "sequential",
            events: [new Event("Request", rawEvent)],
        };
    }
    supportsReverseMapping(eventResponse, response, executionContext) {
        return eventResponse.response instanceof Response;
    }
    reverseMap(eventResponse, response, executionContext) {
        return eventResponse.response;
    }
};
RequestMapper = __decorate([
    moduleScoped(NetworkingModuleKeyname),
    tag(ServiceDefinitionTagEnum.EventMapper),
    injectable()
], RequestMapper);
export { RequestMapper };
