// @pristine-ts/common/dist/lib/esm/enums/service-definition-tag.enum.js
/**
 * This enum defines the default service definition tags. A definition tag is an array of class that can be registered
 * in the container and can be retrieved as an array. For example, if you need to retrieve all the voters, this enum provides
 * an easy way for someone to register a Voter class by simply adding a VOTER token in the container. Then, you can
 * resolve all the voters by requesting the VOTER token and asking the container for a resolveAll.
 */
export var ServiceDefinitionTagEnum;
(function (ServiceDefinitionTagEnum) {
    ServiceDefinitionTagEnum["CurrentChildContainer"] = "CURRENT_CHILD_CONTAINER";
    ServiceDefinitionTagEnum["EventHandler"] = "EventHandlerInterface";
    ServiceDefinitionTagEnum["EventInterceptor"] = "EventInterceptorInterface";
    ServiceDefinitionTagEnum["EventListener"] = "EventListenerInterface";
    ServiceDefinitionTagEnum["EventMapper"] = "EventMapperInterface";
    ServiceDefinitionTagEnum["HttpRequestInterceptor"] = "HttpRequestInterceptorInterface";
    ServiceDefinitionTagEnum["HttpResponseInterceptor"] = "HttpResponseInterceptorInterface";
    ServiceDefinitionTagEnum["HttpErrorResponseInterceptor"] = "HttpErrorResponseInterceptorInterface";
    ServiceDefinitionTagEnum["IdentityProvider"] = "IdentityProviderInterface";
    ServiceDefinitionTagEnum["Logger"] = "LoggerInterface";
    ServiceDefinitionTagEnum["MethodParameterDecoratorResolver"] = "METHOD_PARAMETER_DECORATOR_RESOLVER";
    ServiceDefinitionTagEnum["RequestInterceptor"] = "RequestInterceptorInterface";
    ServiceDefinitionTagEnum["ScheduledTask"] = "ScheduledTaskInterface";
    ServiceDefinitionTagEnum["Tracer"] = "TracerInterface";
    ServiceDefinitionTagEnum["Voter"] = "VoterInterface";
})(ServiceDefinitionTagEnum || (ServiceDefinitionTagEnum = {}));
