// @pristine-ts/core/dist/lib/cjs/core.module.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreModule = void 0;
const core_module_keyname_1 = require("./core.module.keyname");
const telemetry_1 = require("@pristine-ts/telemetry");
const configuration_1 = require("@pristine-ts/configuration");
const logging_1 = require("@pristine-ts/logging");
__exportStar(require("./kernel"), exports);
__exportStar(require("./dispatchers/dispatchers"), exports);
__exportStar(require("./enums/enums"), exports);
__exportStar(require("./errors/errors"), exports);
__exportStar(require("./handlers/handlers"), exports);
__exportStar(require("./interceptors/interceptors"), exports);
__exportStar(require("./interfaces/interfaces"), exports);
__exportStar(require("./listeners/listeners"), exports);
__exportStar(require("./mappers/mappers"), exports);
__exportStar(require("./models/models"), exports);
__exportStar(require("./pipelines/pipelines"), exports);
__exportStar(require("./core.module.keyname"), exports);
exports.CoreModule = {
    keyname: core_module_keyname_1.CoreModuleKeyname,
    importModules: [
        configuration_1.ConfigurationModule,
        telemetry_1.TelemetryModule,
        logging_1.LoggingModule,
    ],
    providerRegistrations: [],
    configurationDefinitions: [
        {
            parameterName: core_module_keyname_1.CoreModuleKeyname + ".requestBodyConverterActive",
            defaultValue: true,
            isRequired: false,
            defaultResolvers: [
                new configuration_1.EnvironmentVariableResolver("PRISTINE_CORE_REQUEST_BODY_CONVERTER_ACTIVE")
            ]
        },
    ],
};
