// @pristine-ts/telemetry/dist/lib/cjs/enums/span-keyname.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanKeynameEnum = void 0;
/**
 * This enum is for the different span names that are integrated in Pristine.
 */
var SpanKeynameEnum;
(function (SpanKeynameEnum) {
    SpanKeynameEnum["ChildContainerCreation"] = "child.container.creation";
    SpanKeynameEnum["ConfigurationInitialization"] = "configuration.initialization";
    SpanKeynameEnum["ErrorResponseInterceptors"] = "error.response.interceptors";
    SpanKeynameEnum["EventExecution"] = "event.execution";
    SpanKeynameEnum["EventInitialization"] = "event.initialization";
    SpanKeynameEnum["EventPreMappingInterception"] = "event.pre-mapping.interception";
    SpanKeynameEnum["EventPostMappingInterception"] = "event.post-mapping.interception";
    SpanKeynameEnum["EventPreResponseMappingInterception"] = "event.pre-response-mapping.interception";
    SpanKeynameEnum["EventPostResponseMappingInterception"] = "event.post-response-mapping.interception";
    SpanKeynameEnum["EventMapping"] = "event.mapping";
    SpanKeynameEnum["KernelInitialization"] = "kernel.initialization";
    SpanKeynameEnum["ModuleInitialization"] = "module.initialization";
    SpanKeynameEnum["ModuleInitializationImportModules"] = "module.initialization.import.modules";
    SpanKeynameEnum["RequestExecution"] = "request.execution";
    SpanKeynameEnum["RequestInterceptors"] = "request.interceptors";
    SpanKeynameEnum["ResponseInterceptors"] = "response.interceptors";
    SpanKeynameEnum["RootExecution"] = "root.execution";
    SpanKeynameEnum["RouterControllerResolver"] = "router.controller.resolver";
    SpanKeynameEnum["RouterFindMethodRouterNode"] = "router.find.method.router.node";
    SpanKeynameEnum["RouterRequestAuthentication"] = "router.request.authentication";
    SpanKeynameEnum["RouterRequestExecution"] = "router.request.execution";
    SpanKeynameEnum["RouterSetup"] = "router.setup";
})(SpanKeynameEnum = exports.SpanKeynameEnum || (exports.SpanKeynameEnum = {}));
