// @pristine-ts/telemetry/dist/lib/cjs/models/trace.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Trace = void 0;
const uuid_1 = require("uuid");
/**
 * This model represents a trace.
 */
class Trace {
    /**
     * This model represents a trace.
     * @param id The unique id of the trace.
     * @param context The context associated with the trace.
     */
    constructor(id, context) {
        /**
         * The timestamp in milliseconds at which the trace was started.
         */
        this.startDate = Date.now();
        /**
         * The context associated with the trace.
         */
        this.context = {};
        /**
         * Whether or not the trace was ended.
         */
        this.hasEnded = false;
        this.id = id !== null && id !== void 0 ? id : (0, uuid_1.v4)();
        this.context = context !== null && context !== void 0 ? context : {};
    }
    /**
     * This returns the duration of the trace in miliseconds.
     */
    getDuration() {
        var _a;
        return ((_a = this.endDate) !== null && _a !== void 0 ? _a : Date.now()) - this.startDate;
    }
}
exports.Trace = Trace;
