// @pristine-ts/networking/dist/lib/cjs/mappers/request.mapper.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestMapper = void 0;
const common_1 = require("@pristine-ts/common");
const tsyringe_1 = require("tsyringe");
const networking_module_keyname_1 = require("../networking.module.keyname");
const core_1 = require("@pristine-ts/core");
let RequestMapper = class RequestMapper {
    supportsMapping(rawEvent, executionContext) {
        return rawEvent instanceof common_1.Request;
    }
    map(rawEvent, executionContext) {
        return {
            executionOrder: "sequential",
            events: [new core_1.Event("Request", rawEvent)],
        };
    }
    supportsReverseMapping(eventResponse, response, executionContext) {
        return eventResponse.response instanceof common_1.Response;
    }
    reverseMap(eventResponse, response, executionContext) {
        return eventResponse.response;
    }
};
RequestMapper = __decorate([
    (0, common_1.moduleScoped)(networking_module_keyname_1.NetworkingModuleKeyname),
    (0, common_1.tag)(common_1.ServiceDefinitionTagEnum.EventMapper),
    (0, tsyringe_1.injectable)()
], RequestMapper);
exports.RequestMapper = RequestMapper;
