// @pristine-ts/telemetry/dist/lib/cjs/tracers/basic.tracer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrayTracer = void 0;
const tsyringe_1 = require("tsyringe");
const common_1 = require("@pristine-ts/common");
const stream_1 = require("stream");
/**
 * We need this to have at least one tracer so the @injectAll(ServiceDefinitionTagEnum.Tracer) does not fail
 * Until there's a fix for: https://github.com/microsoft/tsyringe/issues/63
 */
let XrayTracer = class XrayTracer {
    constructor() {
        this.spanStartedStream = new stream_1.Readable({
            objectMode: true,
            read(size) {
                return true;
            }
        });
        this.spanEndedStream = new stream_1.Readable({
            objectMode: true,
            read(size) {
                return true;
            }
        });
        this.traceStartedStream = new stream_1.Readable({
            objectMode: true,
            read(size) {
                return true;
            }
        });
        this.traceEndedStream = new stream_1.Readable({
            objectMode: true,
            read(size) {
                return true;
            }
        });
    }
};
XrayTracer = __decorate([
    (0, common_1.tag)(common_1.ServiceDefinitionTagEnum.Tracer),
    (0, tsyringe_1.injectable)()
], XrayTracer);
exports.XrayTracer = XrayTracer;
