// @pristine-ts/core/dist/lib/cjs/mappers/default-event.mapper.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultEventMapper = void 0;
const common_1 = require("@pristine-ts/common");
const tsyringe_1 = require("tsyringe");
const core_module_keyname_1 = require("../core.module.keyname");
/**
 * We need a default mapper, else the container will complain when we do injectAll
 */
let DefaultEventMapper = class DefaultEventMapper {
    map(event, executionContext) {
        return {
            events: [],
            executionOrder: 'sequential',
        };
    }
    reverseMap(eventResponse, response, executionContext) {
    }
    supportsMapping(event, executionContext) {
        return false;
    }
    supportsReverseMapping(eventResponse, response, executionContext) {
        return false;
    }
};
DefaultEventMapper = __decorate([
    (0, common_1.moduleScoped)(core_module_keyname_1.CoreModuleKeyname),
    (0, common_1.tag)(common_1.ServiceDefinitionTagEnum.EventMapper),
    (0, tsyringe_1.injectable)()
], DefaultEventMapper);
exports.DefaultEventMapper = DefaultEventMapper;
