// @pristine-ts/express/dist/lib/cjs/mappers/request-event.mapper.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestEventMapper = void 0;
const core_1 = require("@pristine-ts/core");
const common_1 = require("@pristine-ts/common");
const express_module_keyname_1 = require("../express.module.keyname");
const tsyringe_1 = require("tsyringe");
const request_mapper_1 = require("./request.mapper");
const response_mapper_1 = require("./response.mapper");
let RequestEventMapper = class RequestEventMapper {
    constructor(requestMapper, responseMapper) {
        this.requestMapper = requestMapper;
        this.responseMapper = responseMapper;
    }
    supportsMapping(rawEvent, executionContext) {
        // todo be more intelligent and add more conditions
        return executionContext.keyname === core_1.ExecutionContextKeynameEnum.Express;
    }
    map(rawEvent, executionContext) {
        return {
            executionOrder: "sequential",
            events: [
                new core_1.Event("EXPRESS_REQUEST", this.requestMapper.map(rawEvent))
            ]
        };
    }
    reverseMap(eventResponse, response, executionContext) {
        return this.responseMapper.reverseMap(response, executionContext.context.res);
    }
    supportsReverseMapping(eventResponse, response, executionContext) {
        return executionContext.keyname === core_1.ExecutionContextKeynameEnum.Express && response instanceof common_1.Response;
    }
};
RequestEventMapper = __decorate([
    (0, common_1.moduleScoped)(express_module_keyname_1.ExpressModuleKeyname),
    (0, common_1.tag)(common_1.ServiceDefinitionTagEnum.EventMapper),
    (0, tsyringe_1.injectable)(),
    __metadata("design:paramtypes", [request_mapper_1.RequestMapper, response_mapper_1.ResponseMapper])
], RequestEventMapper);
exports.RequestEventMapper = RequestEventMapper;
