// @pristine-ts/aws/dist/lib/esm/enums/s3-event-type.enum.js
/**
 * Event types coming from S3 events
 */
export var S3EventType;
(function (S3EventType) {
    S3EventType["UnknownS3Event"] = "UnknownS3Event";
    S3EventType["ObjectCreatedPut"] = "ObjectCreated:Put";
    S3EventType["ObjectCreatedPost"] = "ObjectCreated:Post";
    S3EventType["ObjectCreatedCopy"] = "ObjectCreated:Copy";
    S3EventType["ObjectCreatedCompleteMultipartUpload"] = "ObjectCreated:CompleteMultipartUpload";
    S3EventType["ObjectRemovedDelete"] = "ObjectRemoved:Delete";
    S3EventType["ObjectRemovedDeleteMarkerCreated"] = "ObjectRemoved:DeleteMarkerCreated";
    S3EventType["ObjectRestorePost"] = "ObjectRestore:Post";
    S3EventType["ObjectRestoreCompleted"] = "ObjectRestore:Completed";
    S3EventType["ReducedRedundancyLostObject"] = "ReducedRedundancyLostObject";
    S3EventType["ReplicationOperationFailedReplication"] = "Replication:OperationFailedReplication";
    S3EventType["ReplicationOperationMissedThreshold"] = "Replication:OperationMissedThreshold";
    S3EventType["ReplicationOperationReplicatedAfterThreshold"] = "Replication:OperationReplicatedAfterThreshold";
    S3EventType["ReplicationOperationNotTracked"] = "Replication:OperationNotTracked";
})(S3EventType || (S3EventType = {}));
