// @pristine-ts/security/dist/lib/cjs/security.module.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityModule = void 0;
const security_module_keyname_1 = require("./security.module.keyname");
const logging_1 = require("@pristine-ts/logging");
const configuration_1 = require("@pristine-ts/configuration");
__exportStar(require("./decorators/decorators"), exports);
__exportStar(require("./enums/enums"), exports);
__exportStar(require("./errors/errors"), exports);
__exportStar(require("./factories/factories"), exports);
__exportStar(require("./guards/guards"), exports);
__exportStar(require("./interfaces/interfaces"), exports);
__exportStar(require("./managers/managers"), exports);
__exportStar(require("./providers/providers"), exports);
__exportStar(require("./security.module.keyname"), exports);
exports.SecurityModule = {
    keyname: security_module_keyname_1.SecurityModuleKeyname,
    importModules: [
        logging_1.LoggingModule,
    ],
    providerRegistrations: [],
    configurationDefinitions: [
        /**
         * The key in the claims of the access token where the roles are defined.
         */
        {
            parameterName: security_module_keyname_1.SecurityModuleKeyname + ".rolesClaimKey",
            isRequired: false,
            defaultValue: "roles",
            defaultResolvers: [
                new configuration_1.EnvironmentVariableResolver("PRISTINE_SECURITY_ROLES_CLAIM_KEY"),
            ]
        }
    ]
};
