// @pristine-ts/configuration/dist/lib/esm/resolvers/boolean.resolver.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ConfigurationResolverError } from "../errors/configuration-resolver.error";
/**
 * This class takes either another resolver or a scalar (boolean, number or string) and returns a boolean value.
 */
export class BooleanResolver {
    constructor(valueOrResolver) {
        this.valueOrResolver = valueOrResolver;
    }
    /**
     * This method takes a string and transforms it into a boolean. This is useful in configurations when the system, for example environment variables, only
     * support string but the configuration expects a boolean.
     * @param value
     * @private
     */
    resolveString(value) {
        const normalizedValue = value.toLowerCase();
        if (normalizedValue === "true" || normalizedValue === "1") {
            return true;
        }
        if (normalizedValue === "false" || normalizedValue === "0") {
            return false;
        }
        throw new ConfigurationResolverError("Cannot convert the string to a boolean.", {
            value
        });
    }
    /**
     * This method takes a number and transforms it into a boolean. This is useful when you receive a number but expect
     * a boolean in the end.
     *
     * @param value
     * @private
     */
    resolveNumber(value) {
        if (value === 0) {
            return false;
        }
        if (value === 1) {
            return true;
        }
        throw new ConfigurationResolverError("Cannot convert the number to a boolean.", value);
    }
    /**
     * This method checks the type of the argument and calls the proper individual method.
     *
     * @param value
     * @private
     */
    resolveValueOrResolver(value) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof value === "boolean") {
                return value;
            }
            if (typeof value === "string") {
                return this.resolveString(value);
            }
            if (typeof value === "number") {
                return this.resolveNumber(value);
            }
            if (typeof value === "object" && typeof value.resolve === "function") {
                return this.resolveValueOrResolver(yield value.resolve());
            }
            throw new ConfigurationResolverError("Cannot resolve the value passed. It isn't of type boolean, string, number or ResolverInterface.", {
                value,
            });
        });
    }
    /**
     * This method resolve the value whether it's a scalar or a Resolver.
     */
    resolve() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.resolveValueOrResolver(this.valueOrResolver);
        });
    }
}
