// @pristine-ts/networking/dist/lib/esm/nodes/method-router.node.js
import { RouterNode } from "./router.node";
import { MethodRouterAddingError } from "../errors/method-router-adding.error";
/**
 * This class represents the Leaf node of the RouteTree.
 * It will always have a parent node that will be a PathRouterNode.
 */
export class MethodRouterNode extends RouterNode {
    /**
     * @param parent The parent node of the current node.
     * @param method The http method of the current method node.
     * @param route The route associated with the current node.
     * @param levelFromRoot The depth level from the root node.
     */
    constructor(parent, method, route, levelFromRoot) {
        super();
        this.method = method;
        this.route = route;
        this.levelFromRoot = levelFromRoot;
        this.parent = parent;
    }
    /**
     * Returns whether or not the node matches.
     * This verification is made by verify the http method matches with the http method of the current node.
     * @param method The http method to compare.
     */
    matches(method) {
        return this.method === method;
    }
    /**
     * A MethodRouterNode is never a catch-all node.
     */
    isCatchAll() {
        return false;
    }
    /**
     * This method adds all the required nodes to match the splitPaths and the method.
     * Since a MethodRouterNode is a leaf node, you can't add anything after it, this method therefore doesn't do anything.
     * @param splitPaths
     * @param method
     * @param route
     */
    add(splitPaths, method, route) {
        throw new MethodRouterAddingError("Impossible to add a child node to a MethodRouterNode.", splitPaths, method, route, this);
    }
    /**
     * This method receives an array of path and recursively calls its children if this node matches
     * the first splitPath. If the node is a MethodRouterNode, it checks to see if the method matches. If yes,
     * it returns itself as the node found. This method should always return a MethodRouterNode. However, Typescript
     * doesn't like these recursive imports so we return the base class
     *
     * @param splitPaths A list of all the parts of the paths spliced at the forward slashes.
     * @param method The http method for which to find a node.
     */
    find(splitPaths, method) {
        if (this.parent && this.parent.isCatchAll()) {
            return this.matches(method) ? this : null;
        }
        return (splitPaths.length === 0 && this.matches(method)) ? this : null;
    }
}
