// @pristine-ts/core/dist/lib/esm/interceptors/event-logging.interceptor.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { moduleScoped, ServiceDefinitionTagEnum, tag } from "@pristine-ts/common";
import { injectable, inject } from "tsyringe";
import { CoreModuleKeyname } from "../core.module.keyname";
/**
 * This class is an interceptor to log the events. It is module scoped to Core module.
 * It is tagged as an EventInterceptor so it can be automatically injected with the all the other EventInterceptors.
 */
let EventLoggingInterceptor = class EventLoggingInterceptor {
    constructor(logHandler) {
        this.logHandler = logHandler;
    }
    preMappingIntercept(event, executionContextInterface) {
        this.logHandler.info("Event just before the EventMapping into an Event object.", {
            event,
            executionContextInterface
        }, CoreModuleKeyname);
        return Promise.resolve(event);
    }
    postMappingIntercept(event) {
        this.logHandler.info("Event just after being mapped into an Event object.", {
            event,
        }, CoreModuleKeyname);
        return Promise.resolve(event);
    }
    preResponseMappingIntercept(eventResponse) {
        this.logHandler.info("Event response just after being dispatched to the Event Listeners.", {
            eventResponse,
        }, CoreModuleKeyname);
        return Promise.resolve(eventResponse);
    }
    postResponseMappingIntercept(eventResponse) {
        this.logHandler.info("Final event response that will be returned.", {
            eventResponse,
        }, CoreModuleKeyname);
        return Promise.resolve(eventResponse);
    }
};
EventLoggingInterceptor = __decorate([
    injectable(),
    moduleScoped(CoreModuleKeyname),
    tag(ServiceDefinitionTagEnum.EventInterceptor),
    __param(0, inject("LogHandlerInterface")),
    __metadata("design:paramtypes", [Object])
], EventLoggingInterceptor);
export { EventLoggingInterceptor };
