// @awslabs-community-fork/dynamodb-expressions/dist/lib/esm/MathematicalExpression.js
import { AttributePath } from "./AttributePath";
import { FunctionExpression } from "./FunctionExpression";
const MATHEMATICAL_EXPRESSION_TAG = 'AmazonDynamoDbMathematicalExpression';
const EXPECTED_TOSTRING = `[object ${MATHEMATICAL_EXPRESSION_TAG}]`;
/**
 * An object representing a DynamoDB function expression.
 */
export class MathematicalExpression {
    constructor(lhs, operator, rhs) {
        this.lhs = lhs;
        this.operator = operator;
        this.rhs = rhs;
        this[Symbol.toStringTag] = MATHEMATICAL_EXPRESSION_TAG;
    }
    serialize(attributes) {
        const safeArgs = [this.lhs, this.rhs].map(arg => {
            if (FunctionExpression.isFunctionExpression(arg)) {
                return arg.serialize(attributes);
            }
            return AttributePath.isAttributePath(arg) || typeof arg === 'string'
                ? attributes.addName(arg)
                : attributes.addValue(arg);
        });
        return `${safeArgs[0]} ${this.operator} ${safeArgs[1]}`;
    }
    /**
     * Evaluate whether the provided value is a MathematicalExpression object.
     */
    static isMathematicalExpression(arg) {
        return arg instanceof MathematicalExpression
            || Object.prototype.toString.call(arg) === EXPECTED_TOSTRING;
    }
}
