// @awslabs-community-fork/dynamodb-expressions/dist/lib/cjs/MathematicalExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MathematicalExpression = void 0;
const AttributePath_1 = require("./AttributePath");
const FunctionExpression_1 = require("./FunctionExpression");
const MATHEMATICAL_EXPRESSION_TAG = 'AmazonDynamoDbMathematicalExpression';
const EXPECTED_TOSTRING = `[object ${MATHEMATICAL_EXPRESSION_TAG}]`;
/**
 * An object representing a DynamoDB function expression.
 */
class MathematicalExpression {
    constructor(lhs, operator, rhs) {
        this.lhs = lhs;
        this.operator = operator;
        this.rhs = rhs;
        this[Symbol.toStringTag] = MATHEMATICAL_EXPRESSION_TAG;
    }
    serialize(attributes) {
        const safeArgs = [this.lhs, this.rhs].map(arg => {
            if (FunctionExpression_1.FunctionExpression.isFunctionExpression(arg)) {
                return arg.serialize(attributes);
            }
            return AttributePath_1.AttributePath.isAttributePath(arg) || typeof arg === 'string'
                ? attributes.addName(arg)
                : attributes.addValue(arg);
        });
        return `${safeArgs[0]} ${this.operator} ${safeArgs[1]}`;
    }
    /**
     * Evaluate whether the provided value is a MathematicalExpression object.
     */
    static isMathematicalExpression(arg) {
        return arg instanceof MathematicalExpression
            || Object.prototype.toString.call(arg) === EXPECTED_TOSTRING;
    }
}
exports.MathematicalExpression = MathematicalExpression;
