// @pristine-ts/common/dist/lib/cjs/models/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Request = void 0;
class Request {
    constructor(httpMethod, url) {
        /**
         * The headers of the request.
         */
        this._headers = {};
        /**
         * The body of the request.
         */
        this.body = {};
        this.httpMethod = httpMethod;
        this.url = url;
    }
    /**
     * This method returns all the headers.
     */
    get headers() {
        return this._headers;
    }
    /**
     * This method sets the headers appropriately.
     *
     * @param headers
     */
    setHeaders(headers) {
        for (const name in headers) {
            if (headers.hasOwnProperty(name) === false) {
                continue;
            }
            this.setHeader(name, headers[name]);
        }
    }
    /**
     * This method sets a header parameter in the Request.
     *
     * @param name The name of the header.
     * @param value The value of the header.
     */
    setHeader(name, value) {
        this.headers[name.toLowerCase()] = value;
    }
    /**
     * This method returns whether or not the header exists in the Request.
     *
     * @param name The name of the header.
     */
    hasHeader(name) {
        return this.headers.hasOwnProperty(name.toLowerCase());
    }
    /**
     * This method returns the header corresponding to the name or undefined if it doesn't exist.
     *
     * @param name The name of the header.
     */
    getHeader(name) {
        return this.headers[name.toLowerCase()];
    }
}
exports.Request = Request;
