// @pristine-ts/common/dist/lib/cjs/models/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = void 0;
class Response {
    constructor() {
        /**
         * The status code of the response. By default we return a 200.
         */
        this.status = 200;
        /**
         * The headers of the response.
         */
        this._headers = {};
    }
    /**
     * This method returns all the headers.
     */
    get headers() {
        return this._headers;
    }
    /**
     * This method sets the headers appropriately.
     *
     * @param headers
     */
    setHeaders(headers) {
        for (const name in headers) {
            if (headers.hasOwnProperty(name) === false) {
                continue;
            }
            this.setHeader(name, headers[name]);
        }
    }
    /**
     * This method sets a header parameter in the Response.
     *
     * @param name The name of the header.
     * @param value The value of the header.
     */
    setHeader(name, value) {
        this.headers[name.toLowerCase()] = value;
    }
    /**
     * This method returns whether or not the header exists in the Response.
     *
     * @param name The name of the header.
     */
    hasHeader(name) {
        return this.headers.hasOwnProperty(name.toLowerCase());
    }
    /**
     * This method returns the header corresponding to the name or undefined if it doesn't exist.
     *
     * @param name The name of the header.
     */
    getHeader(name) {
        return this.headers[name.toLowerCase()];
    }
}
exports.Response = Response;
