// @pristine-ts/networking/dist/lib/cjs/decorators/controller.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.controller = exports.controllerRegistry = void 0;
exports.controllerRegistry = [];
/**
 * The controller decorator can be used on a class to register this class as a controller in the router.
 * @param basePath The base path for all the routes in the controller.
 */
const controller = (basePath) => {
    return (
    /**
     * The constructor of the class
     */
    constructor) => {
        // Verify that the object constructor.prototype["__metadata__"]["controller"][basePath] exists or we create it.
        // This object is a convention defined by Pristine on where to save controller decorator information and is used in the router to retrieve that information.
        if (constructor.prototype.hasOwnProperty("__metadata__") === false) {
            constructor.prototype["__metadata__"] = {};
        }
        if (constructor.prototype["__metadata__"].hasOwnProperty("controller") === false) {
            constructor.prototype["__metadata__"]["controller"] = {};
        }
        // Save the base path.
        constructor.prototype["__metadata__"]["controller"]["basePath"] = basePath;
        // Push the class prototype in the controllerRegistry that is used to instantiate all the controllers for the router.
        exports.controllerRegistry.push(constructor.prototype);
    };
};
exports.controller = controller;
