// @pristine-ts/networking/dist/lib/cjs/decorators/query-parameter.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryParameter = void 0;
/**
 * The queryParameter decorator can be used to inject a specific query parameter of a request in a parameter of a method in a controller.
 * @param name The name of the query parameter to inject.
 */
const queryParameter = (name) => {
    return (
    /**
     * The class on which the decorator is used.
     */
    target, 
    /**
     * The method on which the decorator is used.
     */
    propertyKey, 
    /**
     * The index of the parameter for which the decorator is used.
     */
    parameterIndex) => {
        // Verify that the object target.constructor.prototype["__metadata__"]["methods"][propertyKey]["arguments"] exists or we create it.
        // This object is a convention defined by Pristine on where to save controller method parameter decorator information and is used in the router to retrieve that information.
        if (target.constructor.prototype.hasOwnProperty("__metadata__") === false) {
            target.constructor.prototype["__metadata__"] = {};
        }
        if (target.constructor.prototype["__metadata__"].hasOwnProperty("methods") === false) {
            target.constructor.prototype["__metadata__"]["methods"] = {};
        }
        if (target.constructor.prototype["__metadata__"]["methods"].hasOwnProperty(propertyKey) === false) {
            target.constructor.prototype["__metadata__"]["methods"][propertyKey] = {};
        }
        if (target.constructor.prototype["__metadata__"]["methods"][propertyKey].hasOwnProperty("arguments") === false) {
            target.constructor.prototype["__metadata__"]["methods"][propertyKey]["arguments"] = [];
        }
        // Set the type of method parameter. Each parameter decorator has it's own type.
        // Set also the name of the query parameter to resolve.
        const methodParameter = {
            type: "queryParameter",
            queryParameterName: name,
        };
        // Save the method parameter with the proper parameter index (index of the parameter in the list of parameters of a method).
        target.constructor.prototype["__metadata__"]["methods"][propertyKey]["arguments"][parameterIndex] = methodParameter;
    };
};
exports.queryParameter = queryParameter;
