// @pristine-ts/aws/dist/lib/cjs/clients/s3.client.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3Client = void 0;
const tsyringe_1 = require("tsyringe");
const common_1 = require("@pristine-ts/common");
const aws_module_keyname_1 = require("../aws.module.keyname");
const client_s3_1 = require("@aws-sdk/client-s3");
const s3_presigned_operation_type_enum_1 = require("../enums/s3-presigned-operation-type.enum");
const s3_request_presigner_1 = require("@aws-sdk/s3-request-presigner");
/**
 * The client to use to interact with AWS S3. It is a wrapper around the AWSS3Client of @aws-sdk/client-s3.
 * It is tagged so it can be injected using S3ClientInterface.
 */
let S3Client = class S3Client {
    /**
     * The client to use to interact with AWS S3. It is a wrapper around the AWSS3Client of @aws-sdk/client-s3.
     * @param logHandler The log handler used to output logs.
     * @param region The aws region for which the client will be used.
     */
    constructor(logHandler, region) {
        this.logHandler = logHandler;
        this.region = region;
    }
    /**
     * Returns the instantiated AWSS3Client from the @aws-sdk/client-s3 library
     */
    getClient() {
        var _a;
        return this.client = (_a = this.client) !== null && _a !== void 0 ? _a : new client_s3_1.S3Client({ region: this.region });
    }
    /**
     * Gets an object and all its details from S3.
     * @param bucketName The bucket name where to get the object.
     * @param key The key of the object.
     */
    get(bucketName, key) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logHandler.debug("S3 CLIENT - Getting item", { bucketName, key }, aws_module_keyname_1.AwsModuleKeyname);
            const command = new client_s3_1.GetObjectCommand({
                Bucket: bucketName,
                Key: key,
            });
            try {
                return this.getClient().send(command);
            }
            catch (e) {
                this.logHandler.error("Error getting object from S3", { error: e }, aws_module_keyname_1.AwsModuleKeyname);
                throw e;
            }
        });
    }
    /**
     * Gets an object's body as an array buffer from S3.
     * @param bucketName The bucket name where to get the object.
     * @param key The key of the object.
     */
    getObjectBodyAsArrayBuffer(bucketName, key) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const object = yield this.get(bucketName, key);
                return this.streamToArrayBuffer(object.Body);
            }
            catch (e) {
                this.logHandler.error("Error getting content of object from S3", { error: e }, aws_module_keyname_1.AwsModuleKeyname);
                throw e;
            }
        });
    }
    /**
     * Lists the keys of a bucket.
     * @param bucketName The name of the bucket.
     */
    listKeys(bucketName) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logHandler.debug("S3 CLIENT - Listing bucket keys", { bucketName }, aws_module_keyname_1.AwsModuleKeyname);
            const objects = yield this.listObjects(bucketName);
            return objects.map((object) => object.Key);
        });
    }
    /**
     * Lists the object of a bucket.
     * @param bucketName The name of the bucket.
     */
    listObjects(bucketName) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.logHandler.debug("S3 CLIENT - Listing bucket objects", { bucketName }, aws_module_keyname_1.AwsModuleKeyname);
            const command = new client_s3_1.ListObjectsCommand({
                Bucket: bucketName,
            });
            let objects;
            try {
                objects = yield this.getClient().send(command);
            }
            catch (e) {
                this.logHandler.error("Error listing objects from S3", { error: e }, aws_module_keyname_1.AwsModuleKeyname);
                throw e;
            }
            return (_a = objects.Contents) !== null && _a !== void 0 ? _a : [];
        });
    }
    /**
     * Uploads an object to a bucket of S3.
     * @param bucketName The name of the bucket.
     * @param key The key for the new object.
     * @param data The data to upload.
     * @param contentEncoding The encoding of the data to upload.
     * @param contentType The content type of the data to upload.
     */
    upload(bucketName, key, data, contentEncoding, contentType) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logHandler.debug("S3 CLIENT - Uploading object", { bucketName, key, data, contentEncoding, contentType }, aws_module_keyname_1.AwsModuleKeyname);
            const command = new client_s3_1.PutObjectCommand({
                Bucket: bucketName,
                Key: key,
                Body: data,
                ContentType: contentType,
                ContentEncoding: contentEncoding,
            });
            try {
                yield this.getClient().send(command);
            }
            catch (e) {
                this.logHandler.error("Error putting object in S3", { error: e }, aws_module_keyname_1.AwsModuleKeyname);
                throw e;
            }
        });
    }
    /**
     * Creates a pre signed url to allow a third party to take action on S3.
     * @param bucketName The name of the bucket.
     * @param key The key for the object on which the action will be allowed.
     * @param operation The operation that will be allowed.
     * @param fileName If operation is Get, then a filename can be provided for the name of the file that will be downloaded.
     * @param expiresIn The amount on time in seconds before the pre signed url expires.
     */
    createSignedUrl(bucketName, key, operation, fileName, expiresIn = 300) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logHandler.debug("S3 CLIENT - Creating pre-signed url", { bucketName, key, operation, fileName, expiresIn }, aws_module_keyname_1.AwsModuleKeyname);
            const command = this.getCommandForPresign(operation, bucketName, key, fileName);
            let url;
            try {
                url = yield (0, s3_request_presigner_1.getSignedUrl)(this.getClient(), command, { expiresIn });
            }
            catch (e) {
                this.logHandler.error("Error getting signed url.", { error: e }, aws_module_keyname_1.AwsModuleKeyname);
                throw e;
            }
            return url;
        });
    }
    /**
     * Creates the S3 command for a pre signed url.
     * @param operation The operation that the pre signed url will allow.
     * @param bucketName The name of the bucket.
     * @param key The key for the object on which the action will be allowed.
     * @param fileName If operation is Get, then a filename can be provided for the name of the file that will be downloaded.
     * @private
     */
    getCommandForPresign(operation, bucketName, key, fileName) {
        this.logHandler.debug("S3 CLIENT - Creating command for pre-signed url", { operation, bucketName, key, fileName }, aws_module_keyname_1.AwsModuleKeyname);
        switch (operation) {
            case s3_presigned_operation_type_enum_1.S3PresignedOperationTypeEnum.Get:
                return new client_s3_1.GetObjectCommand({
                    Bucket: bucketName,
                    Key: key,
                    // This specifies the name of the file that will be downloaded rather than using the key.
                    ResponseContentDisposition: fileName ? `attachment; filename=${fileName}` : undefined,
                });
            case s3_presigned_operation_type_enum_1.S3PresignedOperationTypeEnum.Upload:
                return new client_s3_1.PutObjectCommand({
                    Bucket: bucketName,
                    Key: key,
                });
        }
    }
    /**
     * Transforms a stream to an array buffer.
     * @param stream The stream to transform.
     * @private
     */
    streamToArrayBuffer(stream) {
        return __awaiter(this, void 0, void 0, function* () {
            const chunks = [];
            return new Promise((resolve, reject) => {
                stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
                stream.on('error', (err) => reject(err));
                stream.on('end', () => resolve(Buffer.concat(chunks).buffer));
            });
        });
    }
};
S3Client = __decorate([
    (0, common_1.tag)("S3ClientInterface"),
    (0, common_1.moduleScoped)(aws_module_keyname_1.AwsModuleKeyname),
    (0, tsyringe_1.injectable)(),
    __param(0, (0, tsyringe_1.inject)("LogHandlerInterface")),
    __param(1, (0, tsyringe_1.inject)("%pristine.aws.region%")),
    __metadata("design:paramtypes", [Object, String])
], S3Client);
exports.S3Client = S3Client;
