// @pristine-ts/core/dist/lib/esm/mappers/default-event.mapper.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { moduleScoped, ServiceDefinitionTagEnum, tag } from "@pristine-ts/common";
import { injectable } from "tsyringe";
import { CoreModuleKeyname } from "../core.module.keyname";
/**
 * We need a default mapper, else the container will complain when we do injectAll
 */
let DefaultEventMapper = class DefaultEventMapper {
    map(event, executionContext) {
        return {
            events: [],
            executionOrder: 'sequential',
        };
    }
    reverseMap(eventResponse, response, executionContext) {
    }
    supportsMapping(event, executionContext) {
        return false;
    }
    supportsReverseMapping(eventResponse, response, executionContext) {
        return false;
    }
};
DefaultEventMapper = __decorate([
    moduleScoped(CoreModuleKeyname),
    tag(ServiceDefinitionTagEnum.EventMapper),
    injectable()
], DefaultEventMapper);
export { DefaultEventMapper };
